composable_kernel/include/tensor_operation/gridwise_2d_reduction_multiblock_atomic_add.hpp

```diff
@@ -0,0 +1,310 @@
+
+            reducedTiles++;
+        } while(reducedTiles < kBlockTileIterations);
+
+        constexpr auto ReducedDataDesc =
+            make_naive_tensor_descriptor_packed(make_tuple(Number<MThreadSliceSize>{}));
+
+        // Each block executes multiple parallel reductions on the LDS, and by atomic-adding its
+        // reduced output to the global location corresponding to each invariant dimension to get a
+        // consistent reduced result for that invariant dimension. due to the using of vector_load,
+        // each block/thread is involved into multiple invarirant dimensions.
+        static_for<0, MThreadSliceSize, 1>{}([&](auto I) {
+            if constexpr(reorder_thread_cluster)
+            {
+                block_reduce_buf(thread_dim1_cluster_id * MThreadClusterSize +
+                                 thread_dim0_cluster_id) = accuValue_buf[I];
+            }
+            else
+                block_reduce_buf(thread_dim0_cluster_id * KThreadClusterSize +
+                                 thread_dim1_cluster_id) = accuValue_buf[I];
+
+            accuValue_buf(I) = zeroVal;
+
+            __syncthreads();
+
+            blockwise_reduce::Reduce(
+                block_reduce_buf, accuValue_buf(I), thread_dim0_cluster_id, thread_dim1_cluster_id);
+        });
+
+        static_for<0, MThreadSliceSize, 1>{}([&](auto I) {
+            if(thread_dim1_cluster_id == 0)
+            {
+                acc_elementwise_op(accuValue_buf(I), accuValue_buf(I));
+
+                accuValue_buf(I) *= alpha;
+            }
+        });
+
+        if(thread_dim1_cluster_id == 0)
+        {
+            auto threadwise_dst_store =
+                ThreadwiseTensorSliceTransfer_v1r3<AccDataType,
+                                                   OutDataType,
+                                                   decltype(ReducedDataDesc),
+                                                   Out1dDescType,
+                                                   PassThroughOp<AccDataType>,
+                                                   Sequence<MThreadSliceSize>,
+                                                   Sequence<0>,
+                                                   0,
+                                                   OutDstVectorSize,
+                                                   InMemoryDataOperationEnum_t::AtomicAdd,
+                                                   1,
+                                                   true>(
+                    out1dDesc,
+                    make_multi_index(blkgroup_id * M_BlockTileSize +
+                                     thread_dim0_cluster_id * MThreadSliceSize),
+                    PassThroughOp<AccDataType>{});
+
+            threadwise_dst_store.Run(
+                ReducedDataDesc, make_tuple(I0), accuValue_buf, out1dDesc, dst_global_buf);
+        }
+    };
+};
+
+template <index_t BlockSize, typename DataType, typename Global1dBufferDescType>
```

Comment:
could you move this into a device operation? This can be used a lot.
something called:
```DeviceSetTensor```

Reply:
Ok